src/pyflakes/checks.rs

```diff
@@ -31,6 +47,131 @@ pub fn string_dot_format_invalid(literal: &str, location: Range) -> Option<Check
     }
 }
 
+// F522
+pub fn string_dot_format_extra_named_arguments(
+    literal: &str,
+    keywords: &[Keyword],
+    location: Range,
+) -> Option<Check> {
+    if has_star_star_kwargs(keywords) {
+        return None;
+    }
+
+    let keywords = keywords.iter().filter_map(|k| {
+        let KeywordData { arg, .. } = &k.node;
+        arg.clone()
+    });
+
+    match FormatSummary::try_from(literal) {
+        Err(_) => None, // Cannot proceed - should be picked up by F521
+        Ok(summary) => {
+            let missing: Vec<String> = keywords.filter(|k| !summary.keywords.contains(k)).collect();
+
+            if missing.is_empty() {
+                None
+            } else {
+                let message = missing.join(", ");
+                Some(Check::new(
+                    CheckKind::StringDotFormatExtraNamedArguments(message),
+                    location,
+                ))
+            }
+        }
+    }
+}
+
+// F523
+pub fn string_dot_format_extra_positional_arguments(
+    literal: &str,
+    args: &[Expr],
+    location: Range,
+) -> Option<Check> {
+    if has_star_args(args) {
+        return None;
+    }
+
+    match FormatSummary::try_from(literal) {
+        Err(_) => None, // Cannot proceed - should be picked up by F521
+        Ok(summary) => {
+            let missing: Vec<String> = (0..args.len())
+                .filter(|i| !(summary.autos.contains(i) || summary.indexes.contains(i)))
+                .map(|i| i.to_string())
+                .collect();
+
+            if missing.is_empty() {
+                None
+            } else {
+                let message = missing.join(", ");
+                Some(Check::new(
+                    CheckKind::StringDotFormatExtraPositionalArguments(message),
```

Comment:
I'd generally prefer to make `CheckKind::StringDotFormatExtraPositionalArguments` take a `Vec<String>`, and do the joining in `body` (rather than joining here and passing the formatted strings to `CheckKind::StringDotFormatExtraPositionalArguments`.

Reply:
Makes sense - I'll change that